homeassistant/components/transmission/__init__.py

```diff
@@ -55,97 +46,180 @@
+
+    for component in "sensor", "switch":
+        await hass.config_entries.async_forward_entry_unload(entry, component)
+
+    del hass.data[DOMAIN]
+
+    return True
 
-    import transmissionrpc
-    from transmissionrpc.error import TransmissionError
 
+async def get_api(hass, host, port, username=None, password=None):
+    """Get Transmission client."""
     try:
-        api = transmissionrpc.Client(host, port=port, user=username, password=password)
-        api.session_stats()
+        api = await hass.async_add_executor_job(
+            transmissionrpc.Client, host, port, username, password
+        )
+        return api
+
     except TransmissionError as error:
-        if str(error).find("401: Unauthorized"):
-            _LOGGER.error("Credentials for" " Transmission client are not valid")
+        if "401: Unauthorized" in str(error):
+            _LOGGER.error("Credentials for Transmission client are not valid")
+            raise AuthenticationError
+        if "111: Connection refused" in str(error):
+            _LOGGER.error("Connecting to the Transmission client failed")
+            raise CannotConnect
         return False
 
-    tm_data = hass.data[DATA_TRANSMISSION] = TransmissionData(hass, config, api)
 
-    tm_data.update()
-    tm_data.init_torrent_list()
+async def async_populate_options(hass, config_entry):
+    """Populate default options for Transmission Client."""
+    options = {CONF_SCAN_INTERVAL: config_entry.data["options"][CONF_SCAN_INTERVAL]}
+
+    hass.config_entries.async_update_entry(config_entry, options=options)
 
-    def refresh(event_time):
-        """Get the latest data from Transmission."""
-        tm_data.update()
 
-    track_time_interval(hass, refresh, scan_interval)
+class TransmissionClient:
+    """Transmission Client Object."""
 
-    def add_torrent(service):
-        """Add new torrent to download."""
-        torrent = service.data[ATTR_TORRENT]
-        if torrent.startswith(
-            ("http", "ftp:", "magnet:")
-        ) or hass.config.is_allowed_path(torrent):
-            api.add_torrent(torrent)
-        else:
-            _LOGGER.warning(
-                "Could not add torrent: " "unsupported type or no permission"
+    def __init__(self, hass, config_entry):
+        """Initialize the Transmission RPC API."""
+        self.hass = hass
+        self.config_entry = config_entry
+        self.scan_interval = self.config_entry.options[CONF_SCAN_INTERVAL]
+        self.tm_data = None
+        self.unsub_timer = None
+
+    async def async_setup(self):
+        """Set up the Transmission client."""
+
+        config = {
+            CONF_HOST: self.config_entry.data[CONF_HOST],
+            CONF_PORT: self.config_entry.data[CONF_PORT],
+            CONF_USERNAME: self.config_entry.data.get(CONF_USERNAME),
+            CONF_PASSWORD: self.config_entry.data.get(CONF_PASSWORD),
+        }
+        api = await get_api(self.hass, **config)
```

Comment:
We need to handle potential errors here.

Reply:
You mean errors  raised by `get_api`. So shall I use try / exception and return False in case of exception?

Reply:
Yes! :+1:

Non recoverable errors, like bad credential,  should not retry a new config entry setup but just fail that, by returning False in `async_setup_entry`.